### PR TITLE
app.py support local model

### PR DESCRIPTION
**Pull request by AllentDan**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -115,6 +115,45 @@ python3 -m lmdeploy.app {server_ip_addresss}:33337 internlm
 
 For the deployment of other supported models, such as LLaMA, LLaMA-2, vicuna and so on, you can find the guide from [here](docs/en/serving.md)
 
+#### Serving with out docker
```
Comment (lvhan028) [inline README.md:118]: without

File: lmdeploy/app.py [hunk 2]
```diff
@@ -165,5 +171,154 @@ def run(triton_server_addr: str,
...
+        state_chatbot (Sequence): the chatting history
+        llama_chatbot (Chatbot): the instance of a chatbot
+        model_name (str): the name of deployed model
+    """
+    seed = random.getrandbits(64)
+    bot_summarized_response = ''
+    state_chatbot = state_chatbot + [(instruction, None)]
+    instruction = model.get_prompt(instruction, nth_round == 1)
+    if step >= tm_model.session_len:
+        print('WARNING: exceed session max length.'
```
Comment (lvhan028) [inline lmdeploy/app.py:190]: Can this warning be shown on UI?
btw, if it exceeds the limits, server should reject the request

File: README.md (hunk 1 above)
Comment (lvhan028) [inline README.md:118]: Serving with gradio

Comment (lvhan028) [reply README.md:118]: In previous section, the section title can be changed to `Serving with Triton Inference Server`

File: README.md [hunk 3]
```diff
@@ -50,13 +50,14 @@ And the request throughput of TurboMind is 30% higher than vLLM.
 
 ### Installation
 
-Below are quick steps for installation:
+LMDeploy supports python 3.8~3.11. Below are quick steps for installation:
```
Comment (lvhan028) [inline README.md:53]: Install lmdeploy with pip ( python 3.8+) or [from source](./docs/en/build.md)
```shell
pip install lmdeploy
```

File: README.md [hunk 4]
```diff
@@ -92,6 +93,12 @@ python -m lmdeploy.turbomind.chat ./workspace
 > **Note**<br />
 > Tensor parallel is available to perform inference on multiple GPUs. Add `--tp=<num_gpu>` on `chat` to enable runtime TP.
 
+#### Serving with gradio
+
+```shell
+python3 -m lmdeploy.webui.app model_path
```
Comment (lvhan028) [inline README.md:100]: model_path -> ./workspace

Comment (lvhan028) [reply README.md:100]: move   here

File: lmdeploy/serve/gradio/app.py [hunk 5]
```diff
@@ -0,0 +1,334 @@
...
+        share=True,
+        server_port=server_port,
+        server_name=server_name,
+    )
+
+
+def run(model_path_or_server: str,
+        server_name: str = 'localhost',
+        server_port: int = 6006):
+    if ':' in model_path_or_server:
```
Comment (lvhan028) [inline lmdeploy/serve/gradio/app.py:327]: please add docstring to illustrate the launching command clearly

**Review by lvhan028 — APPROVED** [review]

**Comment by irexyc** [comment]
Does it depend on a specific gradio version? I can't run it wht latest gradio version. 
`from gradio.components import *` doesn't import `@ducoment` in latest gradio

After install gradio 3.24.1 and gradio_client 0.0.8, I get this log:
```
(py38) [chenxin@SH-IDC1-10-140-24-141 chenxin]$ python3 -m lmdeploy.serve.gradio.app workspace/
/home/chenxin/miniconda3/envs/py38/lib/python3.8/site-packages/gradio/utils.py:874: UserWarning: Expected 5 arguments for function functools.partial(<function chat_stream_local at 0x7fec0348baf0>, model=<lmdeploy.model.Vicuna object at 0x7fec0348d5b0>, tm_model=<lmdeploy.turbomind.turbomind.TurboMind object at 0x7fec00a00190>, tokenizer=<lmdeploy.turbomind.tokenizer.Tokenizer object at 0x7fec0348d400>, request2instance={}), received 4.
  warnings.warn(
/home/chenxin/miniconda3/envs/py38/lib/python3.8/site-packages/gradio/utils.py:878: UserWarning: Expected at least 5 arguments for function functools.partial(<function chat_stream_local at 0x7fec0348baf0>, model=<lmdeploy.model.Vicuna object at 0x7fec0348d5b0>, tm_model=<lmdeploy.turbomind.turbomind.TurboMind object at 0x7fec00a00190>, tokenizer=<lmdeploy.turbomind.tokenizer.Tokenizer object at 0x7fec0348d400>, request2instance={}), received 4.
  warnings.warn(
/home/chenxin/miniconda3/envs/py38/lib/python3.8/site-packages/gradio/utils.py:874: UserWarning: Expected 5 arguments for function functools.partial(<function reset_local_func at 0x7fec0348bb80>, tm_model=<lmdeploy.turbomind.turbomind.TurboMind object at 0x7fec00a00190>, request2instance={}), received 4.
  warnings.warn(
/home/chenxin/miniconda3/envs/py38/lib/python3.8/site-packages/gradio/utils.py:878: UserWarning: Expected at least 5 arguments for function functools.partial(<function reset_local_func at 0x7fec0348bb80>, tm_model=<lmdeploy.turbomind.turbomind.TurboMind object at 0x7fec00a00190>, request2instance={}), received 4.
  warnings.warn(
server is gonna mount on: http://localhost:6006
Running on local URL:  http://localhost:6006
```

And when I run the model, I get
```
Task exception was never retrieved
future: <Task finished name='bskzc65nxa_0' coro=<Queue.process_events() done, defined at /home/chenxin/miniconda3/envs/py38/lib/python3.8/site-packages/gradio/queueing.py:343> exception=1 validation error for PredictBody
event_id
  Field required [type=missing, input_value={'fn_index': 0, 'data': [...ion_hash': 'bskzc65nxa'}, input_type=dict]
    For further information visit https://errors.pydantic.dev/2.1/v/missing>
Traceback (most recent call last):
  File "/home/chenxin/miniconda3/envs/py38/lib/python3.8/site-packages/gradio/queueing.py", line 347, in process_events
    client_awake = await self.gather_event_data(event)
  File "/home/chenxin/miniconda3/envs/py38/lib/python3.8/site-packages/gradio/queueing.py", line 220, in gather_event_data
    data, client_awake = await self.get_message(event, timeout=receive_timeout)
  File "/home/chenxin/miniconda3/envs/py38/lib/python3.8/site-packages/gradio/queueing.py", line 456, in get_message
    return PredictBody(**data), True
  File "/home/chenxin/miniconda3/envs/py38/lib/python3.8/site-packages/pydantic/main.py", line 159, in __init__
    __pydantic_self__.__pydantic_validator__.validate_python(data, self_instance=__pydantic_self__)
pydantic_core._pydantic_core.ValidationError: 1 validation error for PredictBody
event_id
  Field required [type=missing, input_value={'fn_index': 0, 'data': [...ion_hash': 'bskzc65nxa'}, input_type=dict]
    For further information visit https://errors.pydantic.dev/2.1/v/missing
```

What dependents requires nh3 ? Shoud we add it to requirements?

File: docs/en/build.md [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+  ```
+- install [nccl](https://docs.nvidia.com/deeplearning/nccl/install-guide/index.html), set environment variables:
+  ```shell
+  export NCCL_ROOT_DIR=/path/to/nccl/build
+  export NCCL_LIBRARIES=/path/to/nccl/build/lib
+  ```
+- install rapidjson
+- install openmpi, installing from source is recommended.
+  ```shell
+  wget https://download.open-mpi.org/release/open-mpi/v3.1/openmpi-3.1.0.tar.gz
```
Comment (irexyc) [inline docs/en/build.md:16]:
```suggestion
  wget https://download.open-mpi.org/release/open-mpi/v4.1/openmpi-4.1.5.tar.gz
```

better to use 4.1.5 as it is used in https://github.com/InternLM/lmdeploy/blob/main/docker/Dockerfile

File: docs/zh_cn/build.md [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+  ```
+- 安装 [nccl](https://docs.nvidia.com/deeplearning/nccl/install-guide/index.html),设置环境变量
+  ```shell
+  export NCCL_ROOT_DIR=/path/to/nccl/build
+  export NCCL_LIBRARIES=/path/to/nccl/build/lib
+  ```
+- rapidjson 安装
+- openmpi 安装, 推荐从源码安装:
+  ```shell
+  wget https://download.open-mpi.org/release/open-mpi/v3.1/openmpi-3.1.0.tar.gz
```
Comment (irexyc) [inline docs/zh_cn/build.md:16]:
```suggestion
  wget https://download.open-mpi.org/release/open-mpi/v4.1/openmpi-4.1.5.tar.gz
```
better to use 4.1.5 as it is used in https://github.com/InternLM/lmdeploy/blob/main/docker/Dockerfile

**Review by irexyc — APPROVED** [review]

**Pull request merged**